test/Microsoft.ML.Tests/Transformers/CustomMappingTests.cs

```diff
@@ -121,5 +122,128 @@ public void TestSchemaPropagation()
+                new TextLoader.Column("Label", DataKind.String, 0),
+                new TextLoader.Column("Value", DataKind.Single, 2),
+            });
+            var data = loader.Load(source);
+
+            // We create a temporary environment to instantiate the custom transformer. This is to ensure that we don't need the same
+            // environment for saving and loading.
+            var tempoEnv = new MLContext();
+            var customEst = tempoEnv.Transforms.StatefulCustomMapping<MyStatefulInput, MyState, MyStatefulOutput>(MyStatefulLambda.MyStatefulAction, MyStatefulLambda.MyStateInit, nameof(MyStatefulLambda));
+
+            ML.ComponentCatalog.RegisterAssembly(typeof(MyStatefulLambda).Assembly);
+            TestEstimatorCore(customEst, data);
+            var transformedData = customEst.Fit(data).Transform(data);
+            var outputs = transformedData.GetColumn<bool>(transformedData.Schema[nameof(MyStatefulOutput.FirstAppearance)]);
+            Assert.Equal(10, outputs.Count(output => output));
+
+            Done();
+        }
+
+        [Fact]
+        public void TestCustomFilter()
+        {
+            string dataPath = GetDataPath("breast-cancer.txt");
+            var source = new MultiFileSource(dataPath);
+            var loader = ML.Data.CreateTextLoader(new[] {
+                new TextLoader.Column("Float4", DataKind.Single, 1, 4),
+                new TextLoader.Column("Float1", DataKind.Single, 2),
+            });
+            var data = loader.Load(source);
+
+            var filteredData = ML.Data.CustomFilter<MyInput>(data, input => input.Float1 % 2 == 0);
+            Assert.True(filteredData.GetColumn<float>(filteredData.Schema[nameof(MyInput.Float1)]).All(x => x % 2 == 1));
+        }
+
+        private sealed class MyFilterState
+        {
+            public int Count { get; set; }
+        }
+
+        private sealed class MyFilterInput
+        {
+            public int Counter { get; set; }
+            public int Value { get; set; }
+        }
+
+        [Fact]
+        public void TestStatefulCustomFilter()
+        {
+            var data = ML.Data.LoadFromEnumerable(new[]
+            {
+                new MyFilterInput() { Counter = 0, Value = 1 },
+                new MyFilterInput() { Counter = 1, Value = 1 },
+                new MyFilterInput() { Counter = 2, Value = 2 },
+                new MyFilterInput() { Counter = 3, Value = 0 },
+                new MyFilterInput() { Counter = 4, Value = 2 },
+                new MyFilterInput() { Counter = 5, Value = 4 },
+                new MyFilterInput() { Counter = 6, Value = 1 },
+                new MyFilterInput() { Counter = 7, Value = 1 },
+                new MyFilterInput() { Counter = 8, Value = 2 },
+            });
+
+            var filteredData = ML.Data.StatefulCustomFilter<MyFilterInput, MyFilterState>(data,
+                (input, state) =>
+                {
+                    if (state.Count++ % 2 == 0)
```

Comment:
Will this run on multiple threads? If so, I'm not sure how to get the row number deterministically.

Reply:
The stateful mapping transformer and filter don't use threads, exactly for this reason.

---
In reply to: [397438593](https://github.com/dotnet/machinelearning/pull/4676#discussion_r397438593) [](ancestors = 397438593)